src/weapon/Weapon.ts

```diff
@@ -0,0 +1,198 @@
+import { Game } from '../Game';
+import { Model } from '../Model';
+import { WeaponTint } from './WeaponTint';
+import { WeaponGroup } from './WeaponGroup';
+import { WeaponLivery } from './WeaponLivery';
+import { WeaponLiveryColor } from './WeaponLiveryColor';
+import { WeaponHudStats } from './WeaponHudStats';
+import { enumValues } from '../utils/EnumValues';
+import { Mk2WeaponHash } from './Mk2WeaponHash';
+
+export class Weapon {
+  private readonly owner: Ped;
+  private readonly components: WeaponComponentCollection;
+  private readonly hash: WeaponHash;
+
+  constructor(owner?: Ped, hash?: WeaponHash) {
+    if (owner && hash) {
+      this.owner = owner;
+      this.hash = hash;
+      this.components = new WeaponComponentCollection(this.owner, this);
+    } else {
+      this.hash = WeaponHash.Unarmed;
+    }
+  }
+
+  // public get Components(): WeaponComponentCollection {
+  //   if (!this.components){
+  //     this.components = new WeaponComponentCollection(this.owner, this);
+  //   }
+  //
+  //   return this.components;
+  // }
+
+  public get Components(): WeaponComponentCollection {
+    return this.components;
+  }
+
+  public get Hash(): WeaponHash {
+    return this.hash;
+  }
+
+  public get IsUnarmed(): boolean {
+    return this.hash === WeaponHash.Unarmed;
+  }
+
+  public get IsPresent(): boolean {
+    if (this.IsUnarmed) {
+      return true;
+    }
+
+    return !!HasPedGotWeapon(this.owner.Handle, this.hash, false);
+  }
+
+  public get DisplayName(): string {
+    return Weapon.getDisplayNameFromHash(this.hash) ?? 'WCT_INVALID';
+  }
+
+  public get LocalizedName(): string {
+    return Game.GetGXTEntry(this.DisplayName);
+  }
+
+  public get Model(): Model {
+    return new Model(GetWeapontypeModel(this.hash));
+  }
+
+  public get Tint(): WeaponTint {
+    return GetPedWeaponTintIndex(this.owner.Handle, this.hash);
+  }
+
+  public set Tint(tint: WeaponTint) {
+    SetPedWeaponTintIndex(this.owner.Handle, this.hash, tint);
+  }
+
+  public get Group(): WeaponGroup {
+    return GetWeapontypeGroup(this.hash);
+  }
+
+  public get AmmoType(): AmmoType {
+    return GetPedAmmoTypeFromWeapon(this.owner.Handle, this.hash);
+  }
+
+  public get Ammo(): number {
+    if (this.IsUnarmed) {
+      return 1;
+    }
+
+    if (!this.IsPresent) {
+      return GetPedAmmoByType(this.owner.Handle, this.AmmoType);
+    }
+
+    return GetAmmoInPedWeapon(this.owner.Handle, this.hash);
+  }
+
+  public set Ammo(amount: number) {
+    if (this.IsUnarmed) {
+      return;
+    }
+
+    if (this.IsPresent) {
+      SetPedAmmo(this.owner.Handle, this.hash, amount);
+    } else {
+      GiveWeaponToPed(this.owner.Handle, this.hash, amount, false, true);
+    }
+  }
+
+  public get MaxAmmo(): number {
+    if (this.IsUnarmed) {
+      return 1;
+    }
+
+    // GetMaxAmmo
+    // https://docs.fivem.net/natives/?_0xDC16122C7A20C933
+    // noinspection UnnecessaryLocalVariableJS
+    const amount: number = Citizen.invokeNative(
```

Comment:
Again, please call `GetMaxAmmo` directly :)

Reply:
`GetMaxAmmo` seems have wrong method signature, on which should return [Vector[boolean, number]](https://docs.fivem.net/natives/?_0xDC16122C7A20C933) but defined as number in JavaScript. This might be mistaken and I prefer use invokeNative before actual fix by FiveM
```
// at natives_universal.d.ts
declare function GetMaxAmmo(ped: number, weaponHash: string | number, ammo: number): number;

// at https://docs.fivem.net/natives/?_0xDC16122C7A20C933
const [retval, ammo]: [boolean, number] = 
	GetMaxAmmo(
		ped: Ped, 
		weaponHash: Hash
	);
```

Reply:
This was fixed in recent [@citizenfx/client](https://www.npmjs.com/package/@citizenfx/client?activeTab=readme), plan to resolve once dependency upgraded.